tokio/src/task/task_set.rs

```diff
@@ -0,0 +1,242 @@
+    ///
+    /// This method ignores any panics in the tasks shutting down. When this call returns, the
+    /// `JoinSet` will be empty.
+    ///
+    /// [`abort_all`]: fn@Self::abort_all
+    /// [`join_one`]: fn@Self::join_one
+    pub async fn shutdown(&mut self) {
+        self.abort_all();
+        while self.join_one().await.transpose().is_some() {}
+    }
+
+    /// Aborts all tasks on this `TaskSet`.
+    ///
+    /// This does not remove the tasks from the `TaskSet`. To wait for the tasks to complete
+    /// cancellation, you should call `join_one` in a loop until the `TaskSet` is empty.
+    pub fn abort_all(&mut self) {
+        self.inner.for_each(|jh| jh.abort());
+    }
+
+    /// Removes all tasks from this `TaskSet` without aborting them.
+    ///
+    /// The tasks removed by this call will continue to run in the background even if the `TaskSet`
+    /// is dropped.
+    pub fn detach_all(&mut self) {
+        self.inner.drain(drop);
+    }
+
+    /// Polls for one of the tasks in the set to complete.
+    ///
+    /// If this returns `Poll::Ready(Ok(Some(_)))` or `Poll::Ready(Err(_))`, then the task that
+    /// completed is removed from the set.
+    ///
+    /// When the method returns `Poll::Pending`, the `Waker` in the provided `Context` is scheduled
+    /// to receive a wakeup when a task in the `TaskSet` completes. Note that on multiple calls to
+    /// `poll_join_one`, only the `Waker` from the `Context` passed to the most recent call is
+    /// scheduled to receive a wakeup.
+    ///
+    /// # Returns
+    ///
+    /// This function returns:
+    ///
+    ///  * `Poll::Pending` if the `TaskSet` is not empty but there is no task whose output is
+    ///     available right now.
+    ///  * `Poll::Ready(Ok(Some(value)))` if one of the tasks in this `TaskSet` has completed. The
+    ///    `value` is the return value of one of the tasks that completed.
+    ///  * `Poll::Ready(Err(err))` if one of the tasks in this `TaskSet` has panicked or been
+    ///     aborted.
+    ///  * `Poll::Ready(Ok(None))` if the `TaskSet` is empty.
+    ///
+    /// Note that this method may return `Poll::Pending` even if one of the tasks has completed.
+    /// This can happen if the [coop budget] is reached.
+    ///
+    /// [coop budget]: crate::task#cooperative-scheduling
+    fn poll_join_one(&mut self, cx: &mut Context<'_>) -> Poll<Result<Option<T>, JoinError>> {
+        // The call to `pop_notified` moves the entry to the `idle` list. It is moved back to
+        // the `notified` list if the waker is notified in the `poll` call below.
+        let mut entry = match self.inner.pop_notified(cx.waker()) {
+            Some(entry) => entry,
+            None => {
+                if self.is_empty() {
+                    return Poll::Ready(Ok(None));
+                } else {
+                    // The waker was set by `pop_notified`.
+                    return Poll::Pending;
+                }
+            }
```

Comment:
nit, take it or leave it: IMO, this is slightly nicer, but I don't have a strong opinion here really
```suggestion
            None if self.is_empty() => return Poll::Ready(Ok(None)),
            // The waker was set by `pop_notified`.
            None => return Poll::Pending,
```

Reply:
This runs into a borrow-check error because `self` is still borrowed by the return value of `pop_notified`.

Reply:
ah, yup. nevermind, then!